Update fields in legacy indexes when removing keys.

ragadeeshu opened the pull request:
Not doing this could potentially lead to incorrect index query results
after removal of an indexed key.

cl: Fix a bug where an entity could be temporarily excluded from some index query results if another key indexed on the same entity were removed.

ragadeeshu committed:
Take the elusive none docvaluestype into account

For explicit indexes

MishaDemianenko commented:
It looks like this PR needs a bit more broad description and test

tinwelint reviewed (changes requested):
Some testing should be added for this.

ragadeeshu committed:
Update fields in legacy indexes when removing keys

Not doing this could potentially lead to incorrect index query results
after removal of an indexed key.

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/index/impl/lucene/legacy/IndexType.java:
```diff
@@ -374,6 +374,7 @@ final void removeFromDocument( Document document, String key, Object value )
         else
         {
             removeFieldsFromDocument( document, key, value );
+            restoreSortFields( document );
```
Is this a fix for something unrelated to the original PR changes?

ragadeeshu replied:
Not unrelated. This is currently the only _known_ cause for the case handled in the first commit. One of the tests in this PR tests for this bug explicitly.

tinwelint commented on community/lucene-index/src/test/java/org/neo4j/index/LegacyIndexTest.java:
```diff
@@ -139,6 +142,114 @@ public void shouldThrowIllegalArgumentChangingTypeOfFieldOnNodeIndex()
     }
 
     @Test
+    public void shouldBeAbleToAddNodesAfterRemovalOfKey()
```
This test is green w/o your changes, is that expected? How do we know that your changes solves the issue?

ragadeeshu replied:
This test is made green by the first commit in this PR.

tinwelint replied:
Missed the change to the db rule. All good here.

tinwelint commented on community/lucene-index/src/test/java/org/neo4j/index/LegacyIndexTest.java:
```diff
@@ -139,6 +142,114 @@ public void shouldThrowIllegalArgumentChangingTypeOfFieldOnNodeIndex()
+        {
+            Node node = db.createNode();
+            Index<Node> nodeIndex = db.index().forNodes( indexName );
+            nodeIndex.add( node, "key", "hej" );
```
did you mean to use `keydelete` here as key?

ragadeeshu replied:
No, I want to verify the that the indextype of `key` is correct.
